modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -422,65 +422,133 @@ public PathElement[] getCaretShape(int offset, boolean isLeading,
                             textWidthPrevLine += r.getLength();
                         }
                     }
                 }
-                int prevNodeLength = 0;
-                boolean isPrevNodeExcluded = false;
-                for (TextRun r: runs) {
-                    if (!r.getTextSpan().getText().equals(text) || (r.getStart() < textRunStart && r.getTextSpan().getText().equals(text))) {
-                        prevNodeLength += r.getWidth();
-                        continue;
+
+                if (isMirrored) {
+                    for (TextRun r: runs) {
+                        if (r.getStart() != curRunStart && r.getTextSpan().getText().equals(text)) {
+                            isMultiRunText = true;
+                        }
                     }
-                    if (r.getTextSpan() != null && r.getTextSpan().getText().equals(text)) {
-                        BaseBounds textBounds = new BoxBounds();
-                        getBounds(r.getTextSpan(), textBounds);
-                        if (textBounds.getMinX() == 0 && !isPrevNodeExcluded) {
-                            x -= prevNodeLength;
-                            isPrevNodeExcluded = true;
+
+                    for (int i = 0; i < runs.length; i++) {
+                        run = runs[i];
+                        if (run.getStart() != curRunStart && run.getTextSpan().getText().equals(text) && x > run.getWidth()) {
+                            x -= run.getWidth();
+                            continue;
+                        }
+                        if (run.getTextSpan() != null && run.getTextSpan().getText().equals(text)) {
+                            if ((x > run.getWidth() && !isMultiRunText) || textWidthPrevLine > 0) {
+                                BaseBounds textBounds = new BoxBounds();
+                                getBounds(run.getTextSpan(), textBounds);
+                                x -= (run.getLocation().x - textBounds.getMinX());
+                                break;
+                            }
+                            if (x > run.getWidth()) {
+                                x -= run.getWidth();
+                                relIndex += run.getLength();
+                                continue;
+                            }
+                            for (int j = runs.length - 1; j >= 0; j--) {
+                                if (runs[j].getTextSpan().getText().equals(text) && runs[j].getStart() != curRunStart) {
+                                    LTRIndex += runs[j].getLength();
+                                }
+                                if (runs[j].getStart() == curRunStart) {
+                                    break;
+                                }
+                            }
+                            break;
+                        }
+                        // For only English Text embedded in TextFlow in RTL orientation
```

Comment:
this comment seems misleading (by English you mean LTR, right?)
would it be possible to re-phrase, explaining why?  
does it mean the RTL text has been handled by the previous code block and now we are dealing with LTR?

Reply:
Yes, it handles the LTR text present in RTL mode. Rephrased the comment.